### PR TITLE
Remove array allocations routing in common cases

### PR DESCRIPTION
**Pull request by davidfowl**
While looking at some profiles these array allocations were showing up pretty high on the list once low hanging fruit was removed:
- The array of route values after matching
- The array of candidates matched for a route

Using C# 12's inline array feature, we can use stack allocated arrays in 2 cases:

1. When we have < 5 route values.
2. When we have < 5 route candidates (this is usually 1).

TODO: Check for throughput regressions. We have some BDN tests that I need to run.

Allocations:

Before

|Type|Allocations|Bytes|Average Size \(Bytes\)|
|-|-|-|-|
Microsoft.AspNetCore.Routing.Matching.CandidateState\[1\]|10,000|480,000|48|
System.Collections.Generic.KeyValuePair\<System.String, System.Object\>\[\]|10,000|400,000|40|
Microsoft.AspNetCore.Routing.RouteValueDictionary|10,000|400,000|40|

After

|Type|Allocations|Bytes|Average Size \(Bytes\)|
|-|-|-|-|
Microsoft.AspNetCore.Routing.RouteValueDictionary|10,000|1,119,400|111.94|

We've removed 2 heap allocations but increased the size of RVD (there are 4 extra KVP fields now). This was a request with a single route paramter.

### CODE REVIEW

File: src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs [hunk 1]
```diff
@@ -23,14 +23,57 @@ namespace Microsoft.AspNetCore.Routing;
...
+                // still need to fill it with something.
+                end--;
+            }
+        }
+
+        return new RouteValueDictionary
+        {
+            _arrayStorage = new(items[..start]),
+            _count = start,
+        };
```
Comment (stephentoub) [inline src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:74]: Could this instead be:
```C#
int next = 0;
for (int i = 0; i < items.Length; i++)
{
    if (items[i].Key is not null)
    {
        if (next != i) items[next] = items[i];
        next++;
    }
}

return new RouteValueDictionar
{
    _arrayStorage = new(items[..next]),
    _count = next,
}
```
?

File: src/Http/Http.Abstractions/src/Routing/DynamicArray.cs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    [InlineArray(DefaultCapacity)]
+    struct InlineArray<TItem>
+    {
+#pragma warning disable CA1823 // Avoid unused private fields
+#pragma warning disable IDE0044 // Add readonly modifier
+#pragma warning disable IDE0051 // Remove unused private members
+        private TItem _item0;
+#pragma warning restore IDE0051 // Remove unused private members
+#pragma warning restore IDE0044 // Add readonly modifier
+#pragma warning restore CA1823 // Avoid unused private fields
```
[the remark this answers is not in the document]
Comment (mavasani) [reply src/Http/Http.Abstractions/src/Routing/DynamicArray.cs:141]: Can we file issues in both the Roslyn and roslyn-analyzers repo for the same?

File: src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs [hunk 3]
```diff
@@ -23,14 +23,57 @@ namespace Microsoft.AspNetCore.Routing;
...
+    public static RouteValueDictionary FromValues(Span<KeyValuePair<string, object?>> items)
+    {
+        // We need to compress the array by removing non-contiguous items. We
+        // typically have a very small number of items to process. We don't need
+        // to preserve order.
+        var start = 0;
+        var end = items.Length - 1;
+
+        // We walk forwards from the beginning of the array and fill in 'null' slots.
+        // We walk backwards from the end of the array end move items in non-null' slots
```
Comment (DamianEdwards) [inline src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:46]:
```suggestion
        // We walk backwards from the end of the array and move items in non-null' slots
```

File: src/Http/Http.Abstractions/src/Routing/DynamicArray.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Http/Http.Abstractions/src/Routing/DynamicArray.cs:141]: tagging @Cosifne who will take these.

[the remark this answers is not in the document]
Comment (Cosifne) [reply src/Http/Http.Abstractions/src/Routing/DynamicArray.cs:141]: And assigned the issue to me. I am the IDE side buddy to help with this.

File: src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs (hunk 1 above)
Comment (davidfowl) [reply src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:74]: I didn't write the original code, just copied it, but I'll see if this works.

File: src/Http/Http.Abstractions/src/Routing/DynamicArray.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/Http/Http.Abstractions/src/Routing/DynamicArray.cs:143]: I opened https://github.com/dotnet/roslyn/issues/69143 and https://github.com/dotnet/roslyn-analyzers/issues/6789

**Comment by davidfowl** [comment]
Made a part of the change in https://github.com/dotnet/aspnetcore/pull/49579 (the part that doesn't require public APIs)

**Comment by davidfowl** [comment]
To make this change to `RouteValueDictionary` worthwhile, we need to pool or reuse the instances across requests. To be so far this feels like we're shuffling the cost around (moving from 2 references to 1) but not significantly reducing the overall set of allocations.

**Pull request closed without merging**